src/lib/rpmostree.c

```diff
@@ -34,6 +37,69 @@
  * These APIs access generic global state.
  */
 
+#define IPC_FD 3
+
+static GVariant *
+run_ipc (const char *variant_type, char **args, GError **error)
+{
+  g_autoptr(GSubprocessLauncher) launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_STDOUT_SILENCE | G_SUBPROCESS_FLAGS_STDERR_PIPE);
+  int pair[2];
+  if (socketpair (AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, pair) < 0)
+    return (GVariant*)glnx_null_throw_errno_prefix (error, "couldn't create socket pair");
+  glnx_fd_close int my_sock_fd = glnx_steal_fd (&pair[0]);
+  g_subprocess_launcher_take_fd (launcher, pair[1], IPC_FD);
+
+  g_autoptr(GSocket) my_sock = g_socket_new_from_fd (my_sock_fd, error);
+  if (!my_sock)
+    return NULL;
+  my_sock_fd = -1; /* Ownership was transferred */
+  g_autoptr(GPtrArray) full_args = g_ptr_array_new ();
+  g_ptr_array_add (full_args, "rpm-ostree");
+  g_ptr_array_add (full_args, "shlib-backend");
+  for (char **it = args; it && *it; it++)
+    g_ptr_array_add (full_args, *it);
+  g_ptr_array_add (full_args, NULL);
+  g_autoptr(GSubprocess) proc = g_subprocess_launcher_spawnv (launcher, (const char*const*)full_args->pdata, error);
+  if (!proc)
+    return NULL;
+  
+  g_autofree char *stderr = NULL;
+  if (!g_subprocess_communicate_utf8 (proc, NULL, NULL, NULL, &stderr, error))
+    return NULL;
+
+  if (!g_subprocess_get_successful (proc))
+    return glnx_null_throw (error, "Failed to invoke rpm-ostree shlib-backend: %s", stderr);
+  int flags = 0;
+  int nm = 0;
+  GInputVector iv;
+  GUnixFDMessage **mv;
+  char buffer[1024];
+  iv.buffer = buffer;
+  iv.size = 1;
+  gssize r = g_socket_receive_message (my_sock, NULL, &iv, 1,
```

Comment:
This is cribbed from https://gitlab.gnome.org/GNOME/glib/-/blob/master/gio/tests/unix-fd.c FWIW